src/vsr/superblock_free_set.zig

```diff
@@ -96,18 +159,102 @@ pub const FreeSet = struct {
         }
     }
 
-    /// Marks a free block as allocated, and returns the address.
-    /// Returns null if no free block is available.
-    pub fn acquire(set: *FreeSet) ?u64 {
-        const block = blk: {
-            if (set.index.findFirstSet()) |shard| {
-                break :blk set.find_free_block_in_shard(shard) orelse unreachable;
-            } else {
-                return null;
-            }
+    /// Reserve `reserve_count` free blocks. The blocks are not acquired yet.
+    ///
+    /// Invariants:
+    ///
+    ///   - If a reservation is returned, it covers exactly `reserve_count` free blocks, along with
+    ///     any interleaved already-acquired blocks.
+    ///   - Active reservations are exclusive (i.e. disjoint).
+    ///     (A reservation is active until `forfeit()` is called.)
+    ///
+    /// Returns null if there are not enough blocks free and vacant.
+    /// Returns a reservation which can be used with `acquire()`:
+    /// - The caller should consider the returned Reservation as opaque and immutable.
+    /// - Each `reserve()` call which returns a non-null Reservation must correspond to exactly one
+    ///   `forfeit()` call.
+    pub fn reserve(set: *FreeSet, reserve_count: usize) ?Reservation {
+        assert(set.reservation_state == .reserving);
+        assert(reserve_count > 0);
+
+        var shard_start = find_first_set_bit(
+            set.index,
+            @divFloor(set.reservation_blocks, shard_size),
+            set.index.bit_length,
+        ) orelse return null;
+
+        var block = std.math.max(shard_start * shard_size, set.reservation_blocks);
```

Comment:
It wasn't obvious to me at first that Reservation is not a contiguous range but instead promises that there are enough free blocks in that range. That made this code hard to follow. Maybe a comment along the lines of `Find a range which contains reserve_count free blocks`.

Reply:
:+1: I'll make that more clear. There is a note [here](https://github.com/tigerbeetledb/tigerbeetle/blob/5c123869540ad5c45c88171b869cb0a932f20ff3/src/vsr/superblock_free_set.zig#L166-L167) too:

```zig
///   - If a reservation is returned, it covers exactly `reserve_count` free blocks, along with
///     any interleaved already-acquired blocks.
```